Enable anonymous sender with --anonymous

mlsteele opened the pull request:
Hooks up anonymous sender to the CLI.
Functionality already existed in the engine, this just adds a flag to use it.
keybase encrypt bob -i ~/.sample --hide-self
The ephemeral key is used as the sender key.

mlsteele commented:
r? @oconnor663

oconnor663 commented:
Code looks good. I kind of want to bikeshed the name "hide self", since it kinda sorta sounds like it could mean "don't include self in the recipients list". If the recipients list is visible, it certainly doesn't feel like you're very hidden, even if you're not technically identifiable as the sender.

(@mlsteele pointed out that because we don't randomize recipient order, you're actually *totally* identifiable as the sender, even if you pass `--hide-self`, if you forget to pass `--hide-recipients`. But even without that bug, in cases like a message that's only sent to one other person, you really are effectively identifiable anyway.)

I think an effective "anonymous sender" mode has to avoid letting the sender's keys be visible recipients. We have support for specifying just a subset of keys as not-visible, so it's possible we could plumb something through that to allow *other* recipients to remain visible. But I can imagine situations where even that is fishy. Say Alice, Bob, and Charlie are having a 3-way conversation sending each other many saltpack messages. If a message in that thread includes Bob and Charlie as visible recipients, and a single anonymous recipient, it's pretty clear to everyone that it was sent by Alice. I think that if the sender wants to be anonymous, we pretty much have to anonymize all the recipients too. So maybe we should just say `--anonymous-sender` implies `--hide-recipients` too?

(Recently added node in our saltpack TODO list: If we ever get around to forward-compatible shared device encryption keys, the good-error-messages benefits of visible recipients mostly goes away. We could consider removing visible recipients entirely from a future version of the spec.)

mlsteele commented:
Alright, replaced `--hide-self` with `--anonymous` which implies `--hide-recipients`.

Also added TestSaltpackEncryptAnonymous to make sure the recipients keys are hidden and sender is marked anonymous.

oconnor663 commented:
Nice! LGTM.
